FactoryTest/FactoryTest.ino

```diff
@@ -0,0 +1,342 @@
+    int increment = firstIndex < lastIndex ? 1 : -1;
+    int i = firstIndex;
+    do {
+        digitalWrite(LEDpins[i], LOW);
+        delay(duration);
+        digitalWrite(LEDpins[i], HIGH);
+        if (i == lastIndex) break;
+        i += increment;
+    } while (1);
+}
+
+// Give a sign to the user that an exercise has started 
+void startExercise() {
+    exerciseEachLED(0, numLEDs - 1, 50);
+}
+
+// Give a sign to the user that an exercise has ended 
+void endExercise() {
+    exerciseEachLED(numLEDs - 1, 0, 50);
+}
+
+
+/********************************************************************
+ * Temporary debug code, because we don't yet have a usable serial port or debugger.
+ ********************************************************************/
+
+void writeHexDigitToLights(int hexDigit) {
+    allLEDsOff();
+
+    // Turn on the leftmost light, to show that we're displaying a number
+    digitalWrite(LEDpins[0], LOW);
+
+    // Turn on the lights corresponding to the bits of hexDigit
+    digitalWrite(LEDpins[3], (hexDigit & 8) ? LOW : HIGH);
+    digitalWrite(LEDpins[4], (hexDigit & 4) ? LOW : HIGH);
+    digitalWrite(LEDpins[5], (hexDigit & 2) ? LOW : HIGH);
+    digitalWrite(LEDpins[6], (hexDigit & 1) ? LOW : HIGH);
+
+    delay(DELAY_1Sec);
+    allLEDsOff();
+}
+
+// Write a 16-bit number to the LEDs, in hex. It looks like this ...
+//    all LEDs quickly flash, left to right
+//    high-order hex digit, in binary
+//    next hex digit, in binary
+//    next hex digit, in binary
+//    low-order hex digit, in binary
+//    all LEDs quickly flash, right to left
+void writeNumberToLights(uint16_t number) {
+    exerciseEachLED(0, numLEDs - 1, 50);
+
+    writeHexDigitToLights((number >> 12) & 0xf);
+    delay(DELAY_500ms);
+
+    writeHexDigitToLights((number >> 8) & 0xf);
+    delay(DELAY_500ms);
+
+    writeHexDigitToLights((number >> 4) & 0xf);
+    delay(DELAY_500ms);
+
+    writeHexDigitToLights((number) & 0xf);
+    delay(DELAY_500ms);
+
+    exerciseEachLED(numLEDs - 1, 0, 50);
+}
+
+/********************************************************************
+ * Buttons
+ ********************************************************************/
+
+// The ButtonDebounce object polls the pin, and calls a callback when the pin value changes. There is one ButtonDebounce object per button.
+ButtonDebounce buttonFanUp(FAN_UP_PIN, DELAY_100ms);
+ButtonDebounce buttonFanDown(FAN_DOWN_PIN, DELAY_100ms);
+
+/********************************************************************
+ * Fan
+ ********************************************************************/
+
+// How many milliseconds should there be between readings of the fan speed. A smaller value will update
+// more often, while a higher value will give more accurate and smooth readings.
+const int FAN_SPEED_READING_INTERVAL = 1000;
+
+FanController fanController(FAN_RPM_PIN, FAN_SPEED_READING_INTERVAL, FAN_PWM_PIN);
+
+const int FAN_DUTYCYCLE_MINIMUM = 0;
```

Comment:
Personally I put all constants at the top of the file but willing to be persuaded.

Reply:
Done